google/cloud/pubsub/samples/samples.cc

```diff
@@ -518,27 +518,29 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
   using google::cloud::future;
   using google::cloud::StatusOr;
   [](pubsub::Subscriber subscriber, pubsub::Subscription const& subscription) {
-    std::atomic<int> count{0};
-    google::cloud::promise<void> received_message;
+    std::mutex mu;
+    std::condition_variable cv;
+    bool done = false;
+    int message_count = 0;
     auto result = subscriber.Subscribe(
         subscription, [&](pubsub::Message const& m, pubsub::AckHandler h) {
           std::cout << "Received message " << m << "\n";
-          // Signal the waiting thread to detach the subscription after the
-          // first message.
-          if (count.load() == 0) received_message.set_value();
-          ++count;
-          // Only then ack the message, preventing any new requests from being
-          // issued.
+          std::unique_lock<std::mutex> lk(mu);
+          ++message_count;
+          done = true;
+          cv.notify_one();
```

Comment:
Does it make any difference to notify outside the lock?

Reply:
Don't think it does in practical terms, but changed it anyways to release the mutex first.  Are you worried about the waiting thread waking up only to find the mutex is still locked?

Reply:
Well, turns out that was a bad idea.  If things happen at just the right order releasing the mutex first gives the main thread a chance to check the predicate, skip waiting on the condition variable, and exit the function, destroying the condition variable before it is signaled.

https://source.cloud.google.com/results/invocations/8f4a7072-5bd9-48e1-b7b1-2d6825a4b0fd/targets/cloud-cpp%2Fgithub%2Fgoogle-cloud-cpp%2Fmaster%2Fdocker%2Ftsan-presubmit/log

Reply:
> Are you worried about the waiting thread waking up only to find the mutex is still locked?

That's part of it, but more so that the need to do that usually indicates that something else is awry.

For example, needing to hold the lock while notifying implies that the condition variable is protected by the same mutex which protects its condition ... that the mutex is being used as a synchronization device rather than a mutual-exclusion device.

In particular, the condition variable should never be dead while still referenced by the subscription callback if the subscription is properly terminated before the condition variable is destroyed.  A C++ reference to a destructed object, while perhaps benign, should be avoided.

So, I guess my question is about the semantics of `session.cancel(); session.get()`.  If that doesn't guarantee that we're "after" all subscription callbacks, then I think it should.  If it does (and the "no more callbacks can happen after this point" comments seem to imply that) then I don't understand why we need notify under the lock.

Reply:
> So, I guess my question is about the semantics of `session.cancel(); session.get()`. If that doesn't guarantee that we're "after" all subscription callbacks,

To be accurate, it should guarantee that (a) no more callbacks will start, and (b) all callbacks have invoked `nack()` or `ack()` on the handler.  We really don't want to care about the callback returning because we need to support callbacks that start asynchronous work.

> then I think it should.

I agree.

> If it does (and the "no more callbacks can happen after this point" comments seem to imply that)

That is the intent, but maybe there is an underlying bug.

> then I don't understand why we need notify under the lock.

Good question, let me investigate a bit more.